Nessie support for core

rymurr opened the pull request:
As per the mailing list [annoucenment](https://lists.apache.org/thread.html/r82ca07231110a410069bc9609b286ae08bce6d4e3256196fb73f2a45%40%3Cdev.iceberg.apache.org%3E 
) we would like to contribute integration between Iceberg and Nessie to the Iceberg project.

This PR does the following:
* adds a `NessieCatalog` for core iceberg acid operations
* adds nessie support to the catalog and source interfaces for spark 2 and spark 3
* makes nessie branches and tags addressable for iceberg operations

Please have a look at [Iceberg Spark](https://projectnessie.org/tools/spark/) for a more complete description of Nessie's capabilities with iceberg and [Nessie Features](https://projectnessie.org/features/) for a broader introduction to Nessie.

Note this is currently in draft until a gradle plugin required for testing Nessie has been published.

rymurr commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  private String getWarehouseLocation() {
+    String nessieWarehouseDir = config.get("nessie.warehouse.dir");
```
not sure if this is the best way to get hold of a directory to write tables into. Anyone have any suggestions?

rdblue replied:
In general, I would discourage depending so heavily on Hadoop `Configuration`. Spark and Flink have a way to pass catalog-specific options, which is the best way to configure catalogs.

There is some discussion about this in #1640. I think that catalogs should primarily depend on config passed in a string map, and should only use Hadoop `Configuration` when dependencies (like `HadoopFileIO` or `HiveClient`) require it.

rymurr replied:
I have cleaned this up a bit and tried to follow the pattern you suggested in #1640

rymurr commented on spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceNessieTables.java:
```diff
@@ -0,0 +1,120 @@
+  private NessieClient client;
+  private String branch;
+
+  private Configuration getConfig() throws IOException {
```
The nessie specific tests all modify spark settings and reset the settings at the end. This is to interfere as little as possible w/ the 'normal' iceberg path.

rymurr commented on spark2/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -133,16 +135,29 @@ protected Table findTable(DataSourceOptions options, Configuration conf) {
-    if (path.get().contains("/")) {
-      HadoopTables tables = new HadoopTables(conf);
-      return tables.load(path.get());
+    if (nessie(options.asMap(), conf)) {
```
We identify Nessie as the core catalog/source when there are specific parameters available on the classpath or hadoop config. The idea here is to be fully backwards compatible w/ Hive and Hadoop catalogs.

rdblue replied:
This is probably an area to revisit. Right now, this is written to have minimal changes between 2.4.x and 3.0.x, but I think we will probably want to route all loading from here through a catalog. That will allow us to delegate all of this to Nessie or Hive the same way.

rymurr commented on spark3/src/test/java/org/apache/iceberg/spark/SparkCatalogTestBase.java:
```diff
@@ -83,12 +90,40 @@ public static void dropWarehouse() {
+        this.client = new NessieClient(NessieClient.AuthType.NONE, path, null, null);
+        try {
+          try {
+            this.client.getTreeApi().createEmptyBranch(branch);
```
All Nessie tests are run in their own branch to not interfere with parallel test execution

rymurr commented on spark3/src/test/java/org/apache/iceberg/spark/sql/TestNamespaceSQL.java:
```diff
@@ -56,6 +56,8 @@ public void cleanNamespaces() {
   @Test
   public void testCreateNamespace() {
+    // Nessie namespaces are explicit and do not need to be explicitly managed
+    Assume.assumeFalse(catalogName.endsWith("testnessie"));
```
The concept of a namespace is implicit in Nessie and are therefore not managed through the normal `SupportsNamespaces` interface. We skip tests of this interface when the catalog is a `NessieCatalog`.

rdblue replied:
There are a lot of tests that need this. Should we separate the test cases into different suites?

rymurr commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,340 @@
+/**
+ * Nessie implementation of Iceberg Catalog.
+ */
+public class NessieCatalog extends BaseMetastoreCatalog implements AutoCloseable {
```
We do not extend `SupportsNamespaces` as a Nessie object store supports the concept of namespaces implicitly. A Nessie namespace can be arbitrarily deep but is not explicitly created or stored. Similar to empty folders in git.

rdblue replied:
Should be fine, but I think the trade-off is that you won't be able to list namespaces in a namespace. It will be harder to find the namespaces themselves.

rymurr replied:
Having another look we could add valid impls for `namespaceExists` and `listNamespaces` and do no-op or throw for the others. Then the clients can still navigate namespaces. Thoughts?

rymurr commented:
Opening this up as a reviewable PR to get early feedback.

rdblue commented on build.gradle:
```diff
@@ -891,6 +924,49 @@ project(':iceberg-pig') {
 }
 
+project(':iceberg-nessie') {
+  apply plugin: 'org.projectnessie'
```
What's happening in the Nessie plugin?

rdblue commented on build.gradle:
```diff
@@ -693,6 +709,8 @@ if (jdkVersion == '8') {
       maxHeapSize '2500m'
     }
+    // start and stop quarkus for nessie tests
+    tasks.test.dependsOn("quarkus-start").finalizedBy("quarkus-stop")
```
From the comments in the Iceberg sync, it sounds like this is running a stand-alone Nessie server? Is that something we could handle like the current Hive MetaStore tests, where each test suite creates a new metastore and tears it down after the suite runs?

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,333 @@
+          .stream()
+          .filter(namespacePredicate(namespace))
+          .map(NessieCatalog::toIdentifier)
+          .collect(Collectors.toList());
```
Looks like this will return all tables underneath the given namespace, even if they are nested in other namespaces?

I haven't tested this in spark, does it work as expected?

rymurr replied:
Just checked and the contract is `Return all the identifiers under this namespace.` I took this to mean everything under this and all sub namespaces. If that was not the intention of the method I will fix the predicate.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,333 @@
+          .map(NessieCatalog::toIdentifier)
+          .collect(Collectors.toList());
+    } catch (NessieNotFoundException ex) {
+      throw new RuntimeException("Unable to list tables due to missing ref.", ex);
```
Probably shouldn't use `RuntimeException` here. How about `NotFoundException`?

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,142 @@
+    try {
+      Contents contents = client.getContentsApi().getContents(key, reference.getHash());
+      this.table = contents.unwrap(IcebergTable.class)
+          .orElseThrow(() -> new IllegalStateException("Nessie points to a non-Iceberg object for that path."));
```
Style: Most Iceberg error messages use the form `Cannot <some action>: <reason> (<workaround>)`. Consistency here tends to make at least Iceberg errors more readable and easy to consume.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,142 @@
+          .orElseThrow(() -> new IllegalStateException("Nessie points to a non-Iceberg object for that path."));
+      metadataLocation = table.getMetadataLocation();
+    } catch (NessieNotFoundException ex) {
+      this.table = null;
```
I think this should throw `NoSuchTableException` if the existing metadata is not null because the table was deleted under the reference. You'll probably want to follow the same behavior as the Hive catalog.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,142 @@
+      client.getContentsApi().setContents(key,
+                                          reference.getAsBranch().getName(),
+                                          reference.getHash(),
+                                          String.format("iceberg commit%s", applicationId()),
```
Doesn't look like the format here is quite correct. Missing a space?

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,142 @@
+                                          reference.getHash(),
+                                          String.format("iceberg commit%s", applicationId()),
+                                          newTable);
+    } catch (NessieNotFoundException | NessieConflictException ex) {
```
Is this right for `NotFoundException`? Iceberg will retry failed commits.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,142 @@
+        sparkEnvMethod = sparkEnvClazz.getMethod("get");
+        Class sparkConfClazz = Class.forName("org.apache.spark.SparkConf");
+        sparkConfMethod = sparkEnvClazz.getMethod("conf");
+        appIdMethod = sparkConfClazz.getMethod("getAppId");
```
You can use the `DynFields` helpers to do this a bit more easily.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestParsedTableIdentifier.java:
```diff
@@ -0,0 +1,143 @@
+    ParsedTableIdentifier.getParsedTableIdentifier(path, new HashMap<>());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```
We prefer using `AssertHelpers.assertThrows` so that state after the exception was thrown can be validated. For example, testing `catalog.createTable(invalid)` would not only check `ValidationException` but also verify that the table was not created.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:
```diff
@@ -103,6 +106,10 @@ protected Catalog buildIcebergCatalog(String name, CaseInsensitiveStringMap opti
+      case "nessie":
+        String defaultBranch = options.getOrDefault("nessie_ref", "main");
+        String nessieUrl = options.get("nessie_url");
+        return new NessieCatalog(name, conf, defaultBranch, nessieUrl);
```
Please have a look at #1640, I'd like to standardize how we do this. I do like using `type = nessie`, so we may want to have a lookup that points to the `NessieCatalog` implementation.

rdblue commented on spark3/src/test/java/org/apache/iceberg/spark/sql/TestCreateTableAsSelect.java:
```diff
@@ -50,6 +50,7 @@ public TestCreateTableAsSelect(String catalogName, String implementation, Map<St
   @After
   public void removeTables() {
     sql("DROP TABLE IF EXISTS %s", tableName);
+    sql("DROP TABLE IF EXISTS %s", sourceName);
```
Why was this needed?

rdblue commented:
Thanks, @rymurr! This looks like a great start. I commented in a few places where I noticed some things. Overall, you're going in the right direction.

How do you want to start getting this in? I think it would be good to break it up a bit into smaller commits with a few tests. That way, we can iterate more quickly and we reduce the amount of scope that reviewers need to keep track of. Would it be possible to add just the Nessie module with a few tests and then move on to updating Spark modules?

rymurr commented:
> Thanks, @rymurr! This looks like a great start. I commented in a few places where I noticed some things. Overall, you're going in the right direction.
> 
> How do you want to start getting this in? I think it would be good to break it up a bit into smaller commits with a few tests. That way, we can iterate more quickly and we reduce the amount of scope that reviewers need to keep track of. Would it be possible to add just the Nessie module with a few tests and then move on to updating Spark modules?

Thanks a lot for the feedback @rdblue I will rework this PR to be just the nessie module and will open another for Spark. Will follow the pattern from #1640 for the spark PR.

rymurr commented:
Hey @rdblue I have addressed the bulk of your comments above. Left to do:

1) make decision on namespace support for nessie catalog
2) the meaning of `listTables` in Catalog
3) revised `NessieCatalog` constructor along the lines of your comment in #1640 

We should be publishing 0.2.0 of nessie in the next day or two. Once that is pushed I will update this PR with the new versions and we should have a green build.

rymurr commented:
@rdblue the build should be good to go and all should be ready for review!

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,151 @@
+   *   This is hacky but gets the job done until we can have a more complete commit/audit log.
+   * </p>
+   */
+  private static String applicationId() {
```
For Spark writes, we pass the application ID in through `snapshot.summary()`: https://github.com/apache/iceberg/blob/9af545ed56343b2fa09966166f8da3e7a24100d7/spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchWrite.java#L153

rymurr replied:
Ahh, thanks for that. Much cleaner this way.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+  /**
+   * Try to avoid passing parameters via hadoop config. Dynamic catalog expects Map instead
+   *
+   * todo replace with #1640 style constructor
```
#1640 is in. It uses a no-arg constructor and adds an `initialize(String name, Map<String, String> config)` method to initialize and configure the catalog. I think you should be able to update this now.

I'm hoping that this removes the need to make Spark and Flink depend on the new Nessie and Glue modules. We should make sure we have a test suite we can include here that uses Flink and Spark.

rymurr replied:
This is fixed now, apologies. I fixed the constructor but didn't remove the comment.

I agree that the Catalog portion of Spark3 should work fine now w/o explicitly adding Nessie (or Glue etc). I believe we still need to update the `IcebergSource` to handle custom (Iceberg) catalogs right?

Is the intention to add the new catalogs to the Iceberg shaded jar?

rdblue replied:
> Is the intention to add the new catalogs to the Iceberg shaded jar?

I think it depends. If a catalog pulls in a ton of dependencies and requires updating a lot of the shaded Jar's documentation, then it comes at a high cost. On the other hand, if it uses existing bundled libraries or libraries that can be pulled from the Spark runtime, then it would be easier.

> I believe we still need to update the IcebergSource to handle custom (Iceberg) catalogs right?

Yes, we will need to come up with a way for IcebergSource to work with custom catalogs. Spark has a way for the source to return a catalog and identifier that is used instead of the source directly. That's a much better model, but the problem is that we don't necessarily know what the catalog should be. And, if we redirect to a catalog, we will need to also have a catalog that can load Hadoop tables from a URI. I think this is more of a follow-up.

rymurr replied:
Which mechanism were you thinking for this? Was it `LookupCatalog`? That is rather scala-y but it works. I have something basic working along the lines of `LookupCatalog` but I have some concerns about it. Shall I post a PR as a straw man?

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+  }
+
+  private static String getWarehouseLocation(Configuration config) {
+    String nessieWarehouseDir = config.get(NESSIE_WAREHOUSE_DIR);
```
I don't think that any configuration should come from the Hadoop `Configuration` unless it is used for a Hadoop component, like `HadoopFileIO`. Can you initialize this from the catalog config passed to `initialize`?

rymurr replied:
Hey @rdblue these parameters are initialised in the `initialize` method (I have moved the method up to near the constructor as it was hidden at teh bottom of the class). The initialisation uses the passed `options` where possible and falls back to `Configuration` if not found. This is to make it compatible with Spark2/3 `IcebergSource`. However I am happy to remove this once `IcebergSource` supports custom catalogs (which I hope to tackle next if its not already being worked on)

rdblue replied:
We support the Hive warehouse property because that's how to set it up in Hive. Since we are introducing new configuration for Nessie, I'd really rather not make it so people can depend on using the Hadoop config. Then we will never be able to get rid of it.

rymurr replied:
Ok, removed. The consequence of this is the custom catalog work for `IcebergSource` has to be done before the next release if we want a valid/usable nessie in the release

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    } catch (NessieNotFoundException ex) {
+      if (requestedRef != null) {
+        throw new IllegalArgumentException(String.format("Nessie ref '%s' provided via %s does not exist. " +
+          "This ref must exist before creating a NessieCatalog.", requestedRef, NessieClient.CONF_NESSIE_REF), ex);
```
This error message could easily be incorrect because it doesn't use `CONF_NESSIE_REF` directly. It assumes the caller did.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+      if (s.equals(NessieClient.CONF_NESSIE_URL)) {
+        return url == null ? config.get(s) : url;
+      }
+      return config.get(s);
```
Here as well, I don't think this should pull config from `Configuration`.

rymurr replied:
re-jigged this and above a little bit to make it clear that the hadoop config is only used as a fallback. Hopefully that is more clear. As stated before I hope to remove `Configuration` for anything but IO in a further PR.

rdblue replied:
Why remove it in a follow-up? I'd be concerned about not remembering and then needing to break behavior later.

rymurr replied:
removed later as stated above

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    this.warehouseLocation = inputWarehouseLocation == null ? getWarehouseLocation(config) : inputWarehouseLocation;
+
+    final String requestedRef = ref != null ? ref : config.get(NessieClient.CONF_NESSIE_REF);
+    this.reference = get(requestedRef);
```
Looks like `reference` must never be null, correct?

rymurr replied:
correct. The only way to return from `get` is with a valid reference, otherwise an exception would be thrown. Would you prefer an explicit null check here?

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    ParsedTableIdentifier pti = ParsedTableIdentifier.getParsedTableIdentifier(tableIdentifier, new HashMap<>());
+    UpdateableReference newReference = this.reference;
+    if (pti.getReference() != null) {
+      newReference = get(pti.getReference());
```
I think it would be helpful for `get` to have a better name for uses like this. What about `findReference` or `loadReference`?

rymurr replied:
agreed, fixed

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    ParsedTableIdentifier pti = ParsedTableIdentifier.getParsedTableIdentifier(tableIdentifier, new HashMap<>());
```
How about passing `ImmutableMap.of()` instead of `new HashMap<>()`? That avoids unnecessary object creation. Better yet, what about a version of this that doesn't need to pass a map if there isn't one?

rymurr replied:
agreed, fixed

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+      client.getContentsApi().deleteContents(toKey(identifier), reference.getAsBranch().getName(), reference.getHash(),
+          "no message");
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to drop table as ref is no longer valid.", e);
```
Is this assuming that the `NessieNotFoundException` is referring to the ref because the table was loaded just above? Or is that always used for a ref?

rymurr replied:
You are correct `NessieNotFoundException`  refers to the `ref`. If the table were deleted it would be a conflict exception. This is similar to comparing the error modes of git if you committed to a non-existent ref compared to a merge conflict in the case of changing files in the repo

rymurr replied:
I am not 100% certain that `RuntimeException` is the best avenue here, its definitely an unexpected error and in a sense unrecoverable. There are no exceptions referenced in the javadoc so perhaps a log line and returning `false` is more appropriate?

rdblue replied:
Yeah, I like logging and returning false.

rymurr replied:
done

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to drop table as ref is no longer valid.", e);
+    } catch (NessieConflictException e) {
+      throw new RuntimeException("Failed to drop table as table state needs to be refreshed.");
```
Can't this refresh and complete the operation?

rymurr replied:
correct, now refreshes and tries again

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+      throw new RuntimeException("Failed to drop table as table state needs to be refreshed.");
+    }
+
+    // TODO: purge should be blocked since nessie will clean through other means.
```
In this case, just remove the purge. We do that in our catalog as well because we never delete data as a result of a user action. We garbage collect it later.

rymurr replied:
fixed

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    if (purge && lastMetadata != null) {
+      BaseMetastoreCatalog.dropTableData(ops.io(), lastMetadata);
+    }
+    // TODO: fix this so we don't depend on it in tests.
```
What is this referring to?

rymurr replied:
We don't strictly need to refresh immediately after an operation. This generates an extra call to the backend which typically isn't required. We have to do it because tests do require it. Tests tend to switch branches, perform multiple actions and make several conflicting changes in short order in the same jvm so need explicit refresh. Since the api call isn't expensive we have left the refresh in until a better strategy (or better tests) are devised

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    }
+  }
+
+  private TableIdentifier removeCatalogName(TableIdentifier to) {
```
Util methods seem to be mixed in. I think it may help readability if these were at the bottom, or were static methods in a `NessieUtil` class.

rymurr replied:
good idea, fixed. Static methods have been moved to a util class and private methods have been moved to the bottom with `Override` methods grouped above

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+    reference.refresh();
+  }
+
+  public String getHash() {
```
Is there a more specific name for this? It isn't clear what `catalog.getHash()` should be.

Also, style nit: we avoid using `get` where a more specific verb would add value.

rymurr replied:
changed to `currentHash`. Thoughts? The method returns the current hash as teh catalog understands it.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) throws NoSuchNamespaceException {
```
Should we create a trait just for listing namespaces that are implicit?

rymurr replied:
How do you mean? A new interface that no-ops create, load, drop from `SupportsNamespaces`?

rdblue replied:
Just an interface that omits those. All this needs is to list namespaces, not do anything else.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+  }
+
+  @Override
+  public void initialize(String inputName, Map<String, String> options) {
```
Ignore my comments above, since it looks like you've already added this. Can you merge this with `init` and the constructors?

rymurr replied:
done :-)

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+      this.conf = conf;
+    }
+
+    public Builder setUrl(String url) {
```
Nessie URL? In other places, we configure the connection using `uri`.

rymurr replied:
removed the builder and any URL refs I found

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,431 @@
+      return this;
+    }
+
+    public Builder setRef(String ref) {
```
This is the default ref, right? Tables can override it. I think that would be a better name if you can use this catalog to load other refs.

rymurr replied:
I've deleted the builder as its superfluous after #1640

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,151 @@
+      Snapshot snapshot = current.currentSnapshot();
+      if (snapshot != null) {
+        Map<String, String> summary = snapshot.summary();
+        appId = summary.get("spark.app.id");
```
If an update doesn't create a snapshot, then this method will return the app ID that committed the last snapshot. That may not be correct. Should we create a class in core to hold this information instead? Then we could set it somewhere in Spark and Flink so you'd always have identifiers without needing to resort to reflection?

rymurr replied:
Ahh, good point. Do you mean setting some static state somewhere which holds the current app id? I don't love setting static state, is there a way to tell if the snapshot is the 'correct' snapshot we are looking for?

We were just talking today about how to better handle Nessie commit info, perhaps somethign we could discuss tomorrow on the sync call? cc @jacques-n

rdblue replied:
Maybe we should pass this in through catalog properties?

rdblue replied:
I think we should update our SparkCatalog to pass information into catalogs.

rymurr replied:
cool, we can add it to the `CaseInsensitiveStrngMap` when `SparkCatalog` is initialised? I am not sure of the effect of changing this map from inside `SparkCatalog` though, perhaps passing a copy w/ the extra params?

rdblue replied:
Yeah, we should copy and add it. Let's go forward with this for now and we can fix it in a follow-up.

nastra replied:
I opened https://github.com/apache/iceberg/pull/2664 to address that

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,151 @@
+      io().deleteFile(newMetadataLocation);
+      throw new RuntimeException(String.format("Commit failed: Reference %s does not exist", reference.getName()), ex);
+    } catch (Throwable e) {
+      io().deleteFile(newMetadataLocation);
```
What about doing this delete in a `finally` block if `threw` is false? That's usually a better way than catching `Throwable` and wrapping it in a `RuntimeException`.

rymurr replied:
agreed, fixed

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,151 @@
+  protected void doRefresh() {
+    // break reference with parent (to avoid cross-over refresh)
+    // TODO, confirm this is correct behavior.
+    // reference = reference.copy();
```
This doesn't seem correct to me. I try to maintain single-table state by catalog, so that all uses of a table stay in sync. I think it would make sense to do the same with refs. If you update a branch by refreshing or committing any table, it should also refresh everything that is related to stay in sync. Otherwise, you're left with the problem of not knowing whether two tables with the same ref are in sync.

rymurr replied:
Cool, that makes sense to me. Have removed the comment.

rdblue replied:
@rymurr, do we need to request a refresh for all other tables that use this ref?

rymurr replied:
Yes and no, but the option should at least be there. However I think doing this properly is a bigger question. Whether to refresh depends on the isolation level and the use case. 

Currently if two tables were modified by a different process on the same branch and an iceberg client, who had both those tables cached, committed to only one then the two tables would be out of sync on that iceberg client. But I think that would be true if they were using Hive or Hadoop also, correct? I think rather than addressing this use case in here it would be better to have a wider discussion on how to handle consistency across iceberg catalogs. Thoughts?

rymurr commented:
Thanks a lot for the review @rdblue, I think have addressed all of your comments. I think further discussion on namespace interfaces and commit metadata is warranted.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/ParsedTableIdentifier.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    if (path.contains("#")) {
+      throw new IllegalArgumentException("Currently we don't support referencing by timestamp, # is not allowed in " +
```
We try to use simpler error messages and avoid referring to specific people, like "we" or "you". A good rule of thumb is "Cannot [some action]: [problem[ [(suggestion to fix)]" or "Invalid [something]: [problem]". How about "Invalid table name: # is not allowed (reference by timestamp is not supported)"?

rymurr replied:
agreed, fixed

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/ParsedTableIdentifier.java:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * Convert dataset read/write options to a table and ref/hash.
+   */
+  public static ParsedTableIdentifier getParsedTableIdentifier(TableIdentifier path, Map<String, String> properties) {
```
Can we use a simpler verb, like `parse` or `parseTableIdentifier`? It's wordy to use "get" and then a past tense verb.

rymurr replied:
fixed

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/ParsedTableIdentifier.java:
```diff
@@ -0,0 +1,103 @@
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+public class ParsedTableIdentifier {
```
I'm wondering if there is a more specific name for this class. Maybe something like `TableReference` because it has both an identifier and a ref? Or maybe `NessieIdentifier`? `ParsedTableIdentifier` doesn't really tell me what is different about this as opposed to `TableIdentifier`.

rymurr replied:
I chose `TableReference` in the end. I don't love it but its way better than `ParsedTableIdentifier`. I don't know how to concisely state that its a `TableIdentifier` tied to a specific `Reference`. I have updated the javadoc which I hope clarifies it

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/UpdateableReference.java:
```diff
@@ -0,0 +1,80 @@
+    try {
+      reference = client.getReferenceByName(reference.getName());
+    } catch (NessieNotFoundException e) {
+      throw new NotFoundException(e, "Failure refreshing data, table no longer exists.");
```
Table no longer exists? Or the ref no longer exists?

Also, `NotFoundException` is for files that don't exist. Tables should use `NoSuchTableException`

rymurr replied:
fixed the comment and simplified the exception handling

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:
```diff
@@ -0,0 +1,171 @@
+  public static void create() throws Exception {
+    tempDir = java.nio.file.Files.createTempDirectory(
+        "test",
+        PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rwxrwxrwx")))
```
Most tests use `@Rule TemporaryFolder temp` so that JUnit handles temp lifecycle. I'd recommend doing that here, too.

rymurr replied:
agreed! Bizarre that wasn't used in the first place :-)

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:
```diff
@@ -0,0 +1,171 @@
+    tempDir.delete();
+  }
+
+  static String getContent(NessieCatalog catalog, TableIdentifier tableIdentifier) {
```
This looks like it should have a more specific name because it returns the metadata location for a table.

rymurr replied:
agreed! Bizarre that wasn't used in the first place :-)

rymurr replied:
fixed, contents is an internal nessie name for the objects stored in the Nessie object database. Changed contextualize it to content held for iceberg

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:
```diff
@@ -0,0 +1,171 @@
+      tree.createReference(Branch.of(name, null));
+    } else {
+      tree.createReference(Branch.of(name, hash));
+    }
```
Looks like this could just always pass `hash`.

rymurr replied:
:-) i think you are correct

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestCatalog.java:
```diff
@@ -0,0 +1,53 @@
+    Assert.assertEquals("foo", tables.get(0).namespace().toString());
+    catalog.dropTable(TableIdentifier.of("foo", "baz"));
+    tables = catalog.listTables(Namespace.empty());
+    Assert.assertTrue(tables.isEmpty());
```
I prefer to separate tests into distinct cases. It looks like this combines `testRename` with `testListTables`. Combining test cases into a single method causes failures to prevent other tests from running, which makes the whole suite harder to work with.

rymurr replied:
I prefer splitting as well however in this case the state is important. I have split into 3 tests and introduced a lexicographical ordering annotation. Its a little more cluttered this way but let me know what you think, happy to go either way.

rdblue replied:
Looks good.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:
```diff
@@ -0,0 +1,349 @@
+  }
+
+
+  @SuppressWarnings("VariableDeclarationUsageDistance")
```
What is this suppressing?

rymurr replied:
absolutely nothing! It got carried over from the nessie repo (which has militant checkstyle rules) and I forgot to remove it. Gone now.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:
```diff
@@ -0,0 +1,349 @@
+
+  }
+
+  @Test(expected = CommitFailedException.class)
```
Can you use `AssertHelpers.assertThrows` instead? That way, you can add assertions after the exception to check other things, like that the table has not been modified.

rymurr replied:
:+1:

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,148 @@
+  public FileIO io() {
+    if (fileIO == null) {
+      fileIO = new HadoopFileIO(conf);
+    }
```
We've moved the other implementations to create a FileIO at the catalog level and pass it into TableOperations. You may want to do the same. Also, you'll probably want to update to use the same logic so that the implementation can be overridden dynamically.

rymurr replied:
thanks for the pointer, copied the Hive Catalog now.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/ParsedTableIdentifier.java:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * Convert dataset read/write options to a table and ref/hash.
+   */
+  public static ParsedTableIdentifier getParsedTableIdentifier(String path, Map<String, String> properties) {
```
Does this need properties? The only thing that is used is an optional ref. That could be passed by itself rather than as a map.

rymurr replied:
correct, even further the ref isn't really needed. Thanks for pointing that out, much cleaner now.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:
```diff
@@ -0,0 +1,349 @@
+    try (FileAppender<GenericData.Record> writer = Avro.write(Files.localOutput(location2))
+                                                       .schema(schema)
+                                                       .named("test")
+                                                       .build()) {
```
We typically use a continuation indent of 2 indents / 4 spaces, rather than aligning with the previous method call.

rymurr replied:
:+1:

rymurr commented:
Thanks again @rdblue for the patient code review, hopefully this next round looks better. I will be offline for a few days so will address any comments when I return. Unless @jacques-n or @laurentgo get there first :-)

rymurr commented:
Hey @rdblue, responded to some of your comments. Just a few more things to sort out now :-)

rdblue commented:
I'll take another look. Thanks for the updates!

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+  public void initialize(String inputName, Map<String, String> options) {
+    String fileIOImpl = options.get(CatalogProperties.FILE_IO_IMPL);
+    this.fileIO = fileIOImpl == null ? new HadoopFileIO(config) : CatalogUtil.loadFileIO(fileIOImpl, options, config);
+    this.name = inputName == null ? "nessie" : inputName;
```
It seems strange to me to default this in the catalog rather than in the tests. I would probably use a precondition to validate it isn't null instead.

rymurr replied:
Just to make sure I am looking at the right line of the diff: you mean the name right? I have updated the test to set the catalog name to `nessie` but I was following the other Catalog impls which set the name to eg 'hive' or 'hadoop' if name is unset.

rdblue replied:
If that's what the other catalogs do, then that seems reasonable. I think we eventually want to guarantee that there is always a name, but there are probably cases where we didn't have one. We can look into that separately.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+    this.name = inputName == null ? "nessie" : inputName;
+    this.client = NessieClient.withConfig(options::get);
+
+    this.warehouseLocation = options.get(NESSIE_WAREHOUSE_DIR);
```
Other catalogs use "warehouse" instead of a catalog-specific property. It would be slightly better for consistency to do the same, although it is fine this way. If the `NessieClient` is expecting config like this, it may well be more consistent for Nessie users to always use the full namespaced names.

rymurr replied:
I am not sure I understand what you mean by 'warehouse' here. The original version of this patch used the same as the hadoop catalog (`conf.get("fs.defaultFS") + "/" + ICEBERG_HADOOP_WAREHOUSE_BASE`) but I believe a change was suggested/requested. I personally would much prefer a standard approach than the custom nessie directory. Do you prefer using the Hadoop style one above or the `CatalogProperties.WAREHOUSE_LOCATION` one? 

I have switched in the most recent patch to using `WAREHOUSE_LOCATION`

rdblue replied:
Yeah, I meant using `CatalogProperties.WAREHOUSE_LOCATION`. That's what the other catalogs use for a config property.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+    if (warehouseLocation == null) {
+      throw new IllegalStateException("Parameter nessie.warehouse.dir not set, nessie can't store data.");
+    }
+    final String requestedRef = options.get(NessieClient.CONF_NESSIE_REF);
```
Similar, it may be easier to configure using just `"ref"`.

rymurr replied:
:+1:

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+    if (count >= 5) {
+      logger.error("Cannot drop table: failed after retry (update hash and retry)");
+      return false;
+    }
```
You can replace this with `Tasks`:

```java
Tasks.foreach(identifier)
    .retry(5)
    .stopRetryOn(NessieNotFoundException.class)
    .throwFailureWhenFinished()
    .run(this::dropTableInner)
```

`Tasks` is pretty flexible and allows you to configure exponential backoff, whether to retry on all exceptions but a known list (`stopRetryOn`), whether to retry on specific exceptions (`onlyRetryOn`), and set up callbacks (`onFailure`).

It helps us avoid logic like this everywhere.

rymurr replied:
:+1:

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+      // TODO: fix this so we don't depend on it in tests.
+      refresh();
+    } catch (NessieConflictException e) {
+      throw new CommitFailedException(e, "failed");
```
I'd prefer an error message with more context, like `Failed to rename X to Y`. There's no guarantee that the error message from Nessie has that information. It probably has information about the state of the ref, rather than what was being attempted.

rymurr replied:
yup. I believe I was lazy here b/c that exception is always swallowed by the re-try logic above. Not such a good excuse :-) fixed now

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+    } catch (NessieConflictException e) {
+      throw new CommitFailedException(e, "failed");
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to drop table as ref is no longer valid.", e);
```
This can't happen if the `to` table has already been dropped? Seems like this assumes that the `NessieNotFoundException` refers to the ref.

Maybe we're guaranteed that the ref hasn't changed because this isn't a `NessieConflictException`? If so, a comment would help.

rymurr replied:
If `to` table has already been dropped and our ref wasn't up to date we would get a conflict exception. This is similar to a merge conflict in git: the table has been changed by someone else so there is a conflict exception. The fact that the change is a delete is irrelevant nessie, same error would be thrown if the table had been appended to or the schema were to be changed.

I have opened projectnessie/nessie#477 to track the lack of documentation on the exception classes. In the meantime I have added a comment here

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+    return config;
+  }
+
+  public TreeApi getTreeApi() {
```
Why is this `public`?

rymurr replied:
:thinking: I expect there was a reason at some point. It is only used by tests so I have made it package private for now. The original intention was likely to expose CRUD ops on branches/tags to the user but I think a more structured interface would be better for that

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,141 @@
+    try {
+      reference.refresh();
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to refresh as ref is no longer valid.", e);
```
We may want to let the original exception propagate instead of throwing a `RuntimeException` so that people working with the table can take action if the ref is gone. I'm not really sure what the right thing is here.

rymurr replied:
I would normally agree but `NessieNotFoundException` is a checked exception and I don't want to add that throws clause to the interface. Is there a more specialised unchecked exception that you think would suit?

If it helps the likelihood of this throwing should be very low in real life. Typically a user/service would own their branches and deleting a branch when someone else is actively working on it would be an organisational issue rather than something that should be handled in code. An analogy in git would be someone deleting your branch which is an active PR (without closing the PR).

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,322 @@
+  private IcebergTable table(TableIdentifier tableIdentifier) {
+    try {
+      Contents table = client.getContentsApi().getContents(NessieUtil.toKey(tableIdentifier), reference.getHash());
+      if (table instanceof IcebergTable) {
```
`NessieTableOperations` calls `unwrap` instead of using `instanceof`.

rymurr replied:
nice catch

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,141 @@
+      String fixMsg = reference.isBranch() ?
+          String.format("Update the reference %s and try again", reference.getName()) :
+          String.format("Can't commit to the tag %s", reference.getName());
+      throw new CommitFailedException(ex, "Commit failed: Reference hash is out of date. %s", fixMsg);
```
`CommitFailedException` is used to trigger a table refresh and a retry. Throwing it for `NessieConflictException` seems correct to me, but reading the error messages makes me less sure.

If the ref is a tag, then we don't want to retry because it can't be updated, right? In that case, this should throw some other exception because the table is read-only.

If the ref is a branch, then the ref should be refreshed using the normal retry logic so everything looks good. `doRefresh` will update the ref and then update the table content.

rymurr replied:
Yep! exactly! On line 87 we do `reference.checkMutable()` which will throw `IllegalArgumentException` if its a tag. So by the time we commit we know its a branch. I have cleaned up the erroneous error msg.

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:
```diff
@@ -0,0 +1,141 @@
+          String.format("Can't commit to the tag %s", reference.getName());
+      throw new CommitFailedException(ex, "Commit failed: Reference hash is out of date. %s", fixMsg);
+    } catch (NessieNotFoundException ex) {
+      throw new RuntimeException(String.format("Commit failed: Reference %s does not exist", reference.getName()), ex);
```
Like above, do we know this is the ref because the current ref already loaded the table?

rymurr replied:
I think my comment above covers this as well. The only sensible way this could ever throw is if someone deleted your branch between when you loaded the table and when you tried to commit.

I have made that a bit more explicit in the comment

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java:
```diff
@@ -0,0 +1,94 @@
+    this.reference = reference;
+  }
+
+  public TableIdentifier getTableIdentifier() {
```
Nit: We usually omit `get` from getter method names because it doesn't add value and looks weird in other languages (like Scala and Kotlin).

rymurr replied:
fixed. The pythonista in me doesn't like `get` either :-)

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java:
```diff
@@ -0,0 +1,94 @@
+    if (path.contains("#")) {
+      throw new IllegalArgumentException("Invalid table name:" +
+          " # is not allowed (reference by timestamp is not supported)");
+    }
```
Looks like this check is the only one needed for `#`. It doesn't matter if the identifier also contains `@` and it also doesn't matter if there is more than one `#`.

rymurr replied:
If it doesn't matter much to you I would like to leave it in. I will be adding timestamp support here very soon.

Happy to remove it till then if you prefer too.

rdblue replied:
Sounds fine to me if you're going to update this soon. I just wanted to make sure we didn't have an unnecessarily long method.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestBranchHash.java:
```diff
@@ -0,0 +1,78 @@
+    TableIdentifier foobar = TableIdentifier.of("foo", "bar");
+
+    Table bar = createTable(foobar, 1); // table 1
+    catalog.refresh();
```
Isn't it automatically refreshed because it is shared with the `NessieTableOperations`?

rymurr replied:
References aren't refreshed on commits. So while the ref is shared its out of date everywhere.

This normally isn't a problem but since in this test we are creating a branch off the current state of the 'test-branch-hash' branch we explicitly need the up to date hash. Note: in the `TableOperations` the metadata would be up to date so querying the data would work fine. In normal practice the ref being out of date will only mean we don't get any updates from other users (not the recommended usage pattern) or we would get a merge conflict if we updated the table after someone else has.

rdblue replied:
Hm. The commit is going to trigger a refresh on the table the next time its metadata is accessed. That is usually right away because operations will read the latest metadata to clean up unused files. For Nessie, `doRefresh` is going to update the shared ref. That's why I thought this was redundant. I guess there could be cases where the metadata isn't accessed yet and you do need the refresh.

Not a big deal, so let's not worry about it.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestBranchHash.java:
```diff
@@ -0,0 +1,78 @@
+
+  @Test
+  public void testBasicBranch() throws NessieNotFoundException, NessieConflictException {
+    TableIdentifier foobar = TableIdentifier.of("foo", "bar");
```
Using the table name "foo.bar" is fine, but using `foobar` as a variable name makes the test harder to read because it isn't obvious what `metadataLocation(catalog, foobar)` does exactly. If foobar is a `Namespace`, then it would be the default location for a table, for example. It would be easier to read if this were something more descriptive, like `tableIdent`.

rymurr replied:
agreed, fixed

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestBranchHash.java:
```diff
@@ -0,0 +1,78 @@
+  public void testBasicBranch() throws NessieNotFoundException, NessieConflictException {
+    TableIdentifier foobar = TableIdentifier.of("foo", "bar");
+
+    Table bar = createTable(foobar, 1); // table 1
```
Similar, it would be good to name this `table` for clarity later on in this long test method.

rymurr replied:
:+1:

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestBranchHash.java:
```diff
@@ -0,0 +1,78 @@
+        metadataLocation(catalog, TableIdentifier.of("foo", "bar@test")));
+    // asking for table@branch#hash gives expected regardless of catalog
+    Assert.assertEquals(metadataLocation(catalog, foobar),
+        metadataLocation(catalog, TableIdentifier.of("foo", "bar@" + mainHash)));
```
This is another test I think could be broken into distinct cases with a `@Before` to set up the default branch and table.

rymurr replied:
:+1:

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestCatalogBranch.java:
```diff
@@ -0,0 +1,84 @@
+    NessieCatalog newCatalog = initCatalog("test");
+    String initialMetadataLocation = metadataLocation(newCatalog, foobar);
+    Assert.assertEquals(initialMetadataLocation, metadataLocation(catalog, foobar));
+    Assert.assertEquals(metadataLocation(newCatalog, foobaz), metadataLocation(catalog, foobaz));
```
Context helpers would make this test more readable and would be helpful if it ever fails.

rymurr replied:
ive re-worked this test class. Let me know if thats easier to read

rdblue replied:
Looks a lot better!

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestCatalogBranch.java:
```diff
@@ -0,0 +1,84 @@
+    super("main");
+  }
+
+  @SuppressWarnings("VariableDeclarationUsageDistance")
```
What does this suppress?

rymurr replied:
leftover from strict checkstyle checks, removed

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestCatalogBranch.java:
```diff
@@ -0,0 +1,84 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestCatalogBranch extends BaseTestIceberg {
```
What is the main difference between this suite and the branch suite above? Seems very similar to me, so I think I've missed the point. Smaller test cases that are well named would help.

rymurr replied:
effectively none, removed

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestDefaultCatalogBranch.java:
```diff
@@ -0,0 +1,82 @@
+                               metadataLocation(catalog, foobaz));
+
+    System.out.println(metadataLocation(forwardCatalog, foobar));
+    System.out.println(metadataLocation(catalog, foobar));
```
Nit: println usage.

rymurr replied:
removed

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestDefaultCatalogBranch.java:
```diff
@@ -0,0 +1,82 @@
+
+    catalog.dropTable(foobar);
+    catalog.dropTable(foobaz);
+    tree.deleteBranch("FORWARD", tree.getReferenceByName("FORWARD").getHash());
```
This is another test that doesn't seem very different from the others. Can these be combined into a single suite with good test case names that calls out what is unique about each test?

rymurr replied:
agreed, deleted

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:
```diff
@@ -0,0 +1,348 @@
+      for (GenericData.Record rec : records) {
+        writer.add(rec);
+      }
+    }
```
Could this be refactored to be smaller and use a couple of util functions? Something like `addFile(table, filename)` could help.

rymurr replied:
Can do, I original copied this from one of the catalog tests. I believe those are all package private static methods. Do you know of any test class that has these utils available publicly?

rdblue replied:
I think there's something in Flink, but this is an area where we can generally improve tests.

rdblue commented on nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:
```diff
@@ -0,0 +1,348 @@
+
+    AssertHelpers.assertThrows("Update schema fails with conflict exception, ref not up to date",
+        CommitFailedException.class,
+        () -> icebergTable.updateSchema().addColumn("data", Types.LongType.get()).commit());
```
Do you have a test for concurrent writes to the same table from multiple catalogs? That would be good with two cases: two tables that share the same ref (loaded by the same catalog) and two tables that were loaded by different catalogs and have separate refs.

rymurr replied:
`TestCatalogBranch` does this for the positive case (ie concurrent changes that succeed) and internal nessie for teh negative case. I have added a negative case to `TestCatalogBranch` also. (and renamed the class)

rymurr replied:
Cleaned this up a bit more. We do both tests you suggested, thought it may not be clear its happening. Let me know if the extra comments aren't enough.

rdblue commented:
@rymurr, I did another thorough review with more time looking through the tests. Looking close, but I found a few things.

rymurr commented:
Thanks again for the thorough review @rdblue I have updated w/ your suggestions and rebased. hope I didn't miss anything!

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,336 @@
+ * </p>
+ */
+public class NessieCatalog extends BaseMetastoreCatalog implements AutoCloseable, SupportsNamespaces, Configurable {
+  private static final Logger logger = LoggerFactory.getLogger(NessieCatalog.class);
```
Nit: static final constants should use upper case names, like LOGGER. I'm not sure why style checks didn't catch this.

(Not a blocker)

rymurr replied:
fixed. I was just arguing w/ @jacques-n on this point on Fri ;-) He sided with you.

rymurr replied:
@rdblue just checking the checkstyle,xml in the project and it has a regexp of `(?i)log(ger)?\.(debug|info|warn|error)\(.*` for the logger. eg LOG, log, logger, LOGGER are all acceptable. Want me to switch it to just allow `LOGGER`?

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,336 @@
+           .stopRetryOn(NessieNotFoundException.class)
+           .throwFailureWhenFinished()
+           .run(this::dropTableInner, BaseNessieClientServerException.class);
+      threw = false;
```
Nit: `threw` is no longer needed so this could be simply `return true`. That simplifies the logic at the end of the method to just `return false`.

Up to you whether to change this or not. I know some people strongly prefer only one exit point from a method.

rymurr replied:
fixed. I like your way better too..just a hangover from the refactor

rdblue commented on nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:
```diff
@@ -0,0 +1,336 @@
+    if (warehouseLocation == null) {
+      throw new IllegalStateException("Parameter warehouse not set, nessie can't store data.");
+    }
+    final String requestedRef = options.get(removePrefix.apply(NessieClient.CONF_NESSIE_REF));
```
Did you intend to change this to `"ref"`? Your reply seemed to imply that: https://github.com/apache/iceberg/pull/1587#discussion_r526531219

rymurr replied:
It is just ref now, the `removePrefix` method strips the `nessie.`from the constant in the nessie class. Didn't want to duplicate the constants already in `NessieClient`

rdblue replied:
I missed the `removePrefix` call. Thanks!

rdblue commented:
@rymurr, thanks for all of the test changes, it is now much easier to understand! I don't see any blockers, although it looks like you may have intended to change the `nessie.ref` config to just `ref`. I'm going to go ahead and merge this since it is ready and we can clean that up later if you want to change it.

Thanks for all your hard work getting this ready! I actually quite like the way the Nessie reference works and simplifies assumptions in the catalog and table operations.

rymurr commented:
Thanks for the merge @rdblue!! Super pumped to have this merged. I have the last round of changes ready and will post them with the PR to support timestamps in the table name asap.
